### ISSUE TITLE
Invalid 'Location' header: null

### ISSUE THREAD

**Issue by russ-te2** [cognitect-labs/aws-api#15]
; :op :ListBuckets worked fine, when I went to use :ListObjects I obtained the following; 

```
(aws/invoke s3-client {:op :ListObjects :request {:Bucket "bucket-name"}})
=>
{:cognitect.anomalies/category :cognitect.anomalies/fault,
 :cognitect.anomalies/message "Invalid 'Location' header: null",
 :cognitect.http-client/throwable #error{:cause "Invalid 'Location' header: null",
                                         :via [{:type org.eclipse.jetty.client.HttpResponseException,
                                                :message "Invalid 'Location' header: null",
                                                :at [org.eclipse.jetty.client.HttpRedirector
                                                     redirect
                                                     "HttpRedirector.java"
                                                     166]}],
                                         :trace [[org.eclipse.jetty.client.HttpRedirector
                                                  redirect
                                                  "HttpRedirector.java"
                                                  166]
                                                 [org.eclipse.jetty.client.RedirectProtocolHandler
                                                  onComplete
                                                  "RedirectProtocolHandler.java"
                                                  63]
                                                 [org.eclipse.jetty.client.ResponseNotifier
                                                  notifyComplete
                                                  "ResponseNotifier.java"
                                                  193]
                                                 [org.eclipse.jetty.client.ResponseNotifier
                                                  notifyComplete
                                                  "ResponseNotifier.java"
                                                  185]
                                                 [org.eclipse.jetty.client.HttpReceiver
                                                  terminateResponse
                                                  "HttpReceiver.java"
                                                  454]
                                                 [org.eclipse.jetty.client.HttpReceiver
                                                  responseSuccess
                                                  "HttpReceiver.java"
                                                  401]
                                                 [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                  messageComplete
                                                  "HttpReceiverOverHTTP.java"
                                                  268]
                                                 [org.eclipse.jetty.http.HttpParser parseContent "HttpParser.java" 1470]
                                                 [org.eclipse.jetty.http.HttpParser parseNext "HttpParser.java" 1203]
                                                 [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                  parse
                                                  "HttpReceiverOverHTTP.java"
                                                  158]
                                                 [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                  process
                                                  "HttpReceiverOverHTTP.java"
                                                  119]
                                                 [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                  receive
                                                  "HttpReceiverOverHTTP.java"
                                                  69]
                                                 [org.eclipse.jetty.client.http.HttpChannelOverHTTP
                                                  receive
                                                  "HttpChannelOverHTTP.java"
                                                  90]
                                                 [org.eclipse.jetty.client.http.HttpConnectionOverHTTP
                                                  onFillable
                                                  "HttpConnectionOverHTTP.java"
                                                  113]
                                                 [org.eclipse.jetty.io.AbstractConnection$ReadCallback
                                                  succeeded
                                                  "AbstractConnection.java"
                                                  273]
                                                 [org.eclipse.jetty.io.FillInterest fillable "FillInterest.java" 95]
                                                 [org.eclipse.jetty.io.ssl.SslConnection
                                                  onFillable
                                                  "SslConnection.java"
                                                  197]
                                                 [org.eclipse.jetty.io.AbstractConnection$ReadCallback
                                                  succeeded
                                                  "AbstractConnection.java"
                                                  273]
                                                 [org.eclipse.jetty.io.FillInterest fillable "FillInterest.java" 95]
                                                 [org.eclipse.jetty.io.SelectChannelEndPoint$2
                                                  run
                                                  "SelectChannelEndPoint.java"
                                                  75]
                                                 [org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume
                                                  produceAndRun
                                                  "ExecuteProduceConsume.java"
                                                  213]
                                                 [org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume
                                                  run
                                                  "ExecuteProduceConsume.java"
                                                  147]
                                                 [org.eclipse.jetty.util.thread.QueuedThreadPool
                                                  runJob
                                                  "QueuedThreadPool.java"
                                                  654]
                                                 [org.eclipse.jetty.util.thread.QueuedThreadPool$3
                                                  run
                                                  "QueuedThreadPool.java"
                                                  572]
                                                 [java.lang.Thread run "Thread.java" 748]]},
 :cognitect.http-client/meta {:op :ListObjects, :request {:Bucket "bucket-name"}}}
```

**Comment by dchelimsky** [comment]
I'm not able to reproduce this. What versions of `com.cognitect.aws/api`, `com.cognitect.aws/endpoints`, and `com.cognitect.aws/s3` are you using?

**Comment by russ-te2** [comment]
:dependencies [[org.clojure/clojure "1.10.0-RC3"]
                 [com.cognitect.aws/api "0.8.122"]
                 [com.cognitect.aws/endpoints "1.1.11.457"]
                 [com.cognitect.aws/s3 "661.2.363.0"]]

**Comment by dchelimsky** [comment]
Sorry, but I still can't reproduce it. I tried those deps using both tools.deps and a new leiningen project and in both cases I get data if I own the bucket and permissions issues (reported from AWS, not a client side error) if I don't. What versions of java and leiningen are you using?

**Comment by russ-te2** [comment]
new leiningen project;

```
project aws-clojure "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.10.0-RC3"]
                 [com.cognitect.aws/api "0.8.122"]
                 [com.cognitect.aws/endpoints "1.1.11.457"]
                 [com.cognitect.aws/s3 "661.2.363.0"]]
  :main ^:skip-aot aws-clojure.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})
```

```
./java -version 
openjdk version "1.8.0_152-release"
OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b15)
OpenJDK 64-Bit Server VM (build 25.152-b15, mixed mode)
```
I am seeing; 

```
 #'aws-clojure.core/s3-client
Dec 04, 2018 5:28:52 PM clojure.tools.logging$eval10108$fn__10111 invoke
SEVERE: Error fetching credentials from environment variables.
Dec 04, 2018 5:28:52 PM clojure.tools.logging$eval10108$fn__10111 invoke
SEVERE: Error fetching credentials from system properties.
```

But I'm relying on the `AWS_PROFILE` environment variable which I can confirm works
from the command line with `aws s3 ls s3://bucket/`. Also `cognitect.aws.util/getenv AWS_PROFILE` returns the expected profile string.

**Comment by dchelimsky** [comment]
The "SEVERE: Error fetching credentials from environment variables." is a red herring. It really means it couldn't find the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`, even though you were able to get credentials from the profile. You'd not have been able to `ListBuckets` otherwise. I'll change those to warnings and make them more explicit.

I still can't reproduce this, even with your project file copied as/is. I'm using a different java, but I can't imagine that that's the problem:

```
$ java -version
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)
```

Not sure what to do next. What are you using for your repl?

**Comment by dchelimsky** [comment]
Nothing convenient, sorry. Something to think about adding, but for now your best bet is probably cloning the repo and instrumenting the `cognitect.aws.client/send-request` fn to print it out.

**Comment by dchelimsky** [comment]
I added the http-request to the metadata of the response, even if you get an exception, as long as it's after the request is packaged up. Released in 0.8.142.

**Comment by russ-te2** [comment]
Think I tracked it down, it has something to do with which region the bucket was created in. I could :ListObjects fine in the 'working-bucket'.

`
user=> (aws/invoke s3-client {:op :GetBucketLocation :request {:Bucket working-bucket}})
{}
user=> (aws/invoke s3-client {:op :GetBucketLocation :request {:Bucket broken-bucket}})
{:LocationConstraint "us-west-1"}
`

**Comment by russ-te2** [comment]
wow that's counterintuitive; 

`
(def s3-client (aws/client {:api :s3 :region :us-west-1}))
`
creating the client in the same region as the bucket fixes the issue.

**Comment by dchelimsky** [comment]
Well, at least now I can reproduce it! Thanks for tracking it down.

**Comment by dchelimsky** [comment]
AWS is responding with a redirect with no Location header, and jetty throws an exception: https://github.com/jliu-futurewei/jetty/blob/master/jetty-client/src/main/java/org/eclipse/jetty/client/HttpRedirector.java#L152-L175. I'm not sure how we'll address this yet.

**Comment by dchelimsky** [comment]
Actually, I'm not 100% sure that's correct yet, so let me rephrase: by the time the redirect response gets to the Redirector in jetty, there is no Location header.

**Comment by dchelimsky** [comment]
FYI - I changed the title to make it easier for ppl to find this issue.

**Comment by dchelimsky** [comment]
Looking at AWS CLI and SDK implementations, endpoint resolution for s3 requires special handling.

**Comment by dchelimsky** [comment]
@russ-te2 just FYI that I've done some more research. The first observation is that you should actually make a client in the same region as the bucket (see https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAPI.html and https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html). The second is that we were broken for any bucket outside `us-east-1`, so I just released 0.8.171 with a fix so that you're good if the  client and bucket regions are aligned. I'm leaving this open because it'd be nice to have a better error message than "Invalid 'Location' header: null", but I can't tell you when we'll address that.

**Comment by dchelimsky** [comment]
I updated the README with a troubleshooting section.

**Comment by dchelimsky** [comment]
Good news! We finally got to the root cause here!

* The jetty client defaults to follow redirects
* The cognitect http-client did not tell the jetty client to do anything different
* AWS was sending a 301 with no Location header
* The jetty client threw in the face of a 301 with no Location header

The fix:
* cognitect http-client now exposes control over whether to follow redirects
* aws-api now tells the http-client _not_ to follow redirects

The result:
* As of ~~[0.8.870](https://github.com/cognitect-labs/aws-api/blob/main/CHANGES.md#08670--2023-05-22)~~ in addition to `:status` and the "x-amz-bucket-region" in the header, you'll now get an AWS error message with direction to resend the request to the correct region instead of an exception and a stack trace.
* Obs: we chose not to automatically follow this redirect because doing so would actually be a breaking change since it was never following the redirects before thanks to the error reported in this issue.
